cmd/expiration-mailer/main.go

```diff
@@ -277,6 +270,14 @@ func (m *mailer) processCerts(ctx context.Context, allCerts []core.Certificate)
 	}
 }
 
+// findExpiringCertificates finds certificates that might need an expiration mail, filters them,
+// groups by account, sends mail, and updates their status in the DB so we don't examine them again.
+//
+// Invariant: findExpiringCertificates should examine each certificate at most N times, where
+// N is the number of reminders. For every certificate examined (barring errors), this function
+// should update the lastExpirationNagSent field of certificateStatus, so it does not need to
+// examine the same certificate again on the next go-round. This ensures we make forward progress
+// and don't clog up the window of certificates to be examined.
```

Comment:
I support adding this explicit comment documenting the invariant. However, it is not currently true: if the `SelectCertificate` call on line 345 truly fails, then we will bail out at line 356 and end up querying for some of the same certs again in the next batch.

Personally I'd be in favor of fixing that bug in this same PR so that the invariant is immediately true.

Reply:
Merged #6101, so the invariant will now be true when this branch has master merged into it.